lambda-buffers-codegen/src/LambdaBuffers/Codegen/Rust/Print/InstanceDef.hs

```diff
@@ -0,0 +1,60 @@
+module LambdaBuffers.Codegen.Rust.Print.InstanceDef (printInstanceDef, printTraitBound, collectTyVars, printInstanceContext, printInstanceContext') where
+
+import Control.Lens (view)
+import Data.Foldable (Foldable (toList))
+import Data.Set (Set)
+import Data.Set qualified as Set
+import LambdaBuffers.Codegen.Rust.Print.Syntax qualified as R
+import LambdaBuffers.Codegen.Rust.Print.TyDef (printTyInner)
+import LambdaBuffers.ProtoCompat qualified as PC
+import Prettyprinter (Doc, align, braces, colon, comma, encloseSep, group, hardline, langle, line, rangle, space, (<+>))
+
+{- | `printInstanceDef rsQTraitName ty` return a function that given the printed implementation, creates an entire 'instance <rsQTraitName> <ty> where' clause.
+
+
+```rust
+impl SomeClass for SomeSmallTy where
+  someMethod = <implementation>
+
+impl<A: SomeClass, B: SomeClass, C: SomeClass> SomeClass for SomeTy<A, B, C> {
+  fn some_method = <implementation>
+}
+```
+-}
+printInstanceDef :: R.QTraitName -> PC.Ty -> (Doc ann -> Doc ann)
+printInstanceDef rsQTraitName ty =
+  let headDoc = R.printRsQTraitName rsQTraitName <+> "for" <+> printTyInner ty
+      freeVars = collectTyVars ty
+   in case freeVars of
+        [] -> \implDoc -> "impl" <+> headDoc <+> braces (line <> implDoc)
+        _ -> \implDoc ->
+          "impl" <> printInstanceContext rsQTraitName freeVars
+            <+> headDoc
+            <+> braces (hardline <> space <> space <> implDoc)
+
+printInstanceContext :: R.QTraitName -> [PC.Ty] -> Doc ann
+printInstanceContext rsQTraitName = printInstanceContext' [rsQTraitName]
+
+defaultTraitBounds :: [R.QTraitName]
+defaultTraitBounds = [R.qLibRef R.MkTraitName "std" "clone" "Clone"]
+
+printInstanceContext' :: [R.QTraitName] -> [PC.Ty] -> Doc ann
+printInstanceContext' rsQTraitNames tys =
+  align . group $ encloseSep langle rangle comma [printTraitBound (rsQTraitNames <> defaultTraitBounds) ty | ty <- tys]
+
+printTraitBound :: [R.QTraitName] -> PC.Ty -> Doc ann
+printTraitBound qcns ty =
+  let crefDocs = R.printRsQTraitName <$> qcns
+      tyDoc = printTyInner ty
+   in tyDoc <> colon <+> encloseSep mempty mempty "+" crefDocs
+
+collectTyVars :: PC.Ty -> [PC.Ty]
+collectTyVars = fmap (`PC.withInfoLess` (PC.TyVarI . PC.TyVar)) . toList . collectVars
+
+collectVars :: PC.Ty -> Set (PC.InfoLess PC.VarName)
+collectVars = collectVars' mempty
+
+collectVars' :: Set (PC.InfoLess PC.VarName) -> PC.Ty -> Set (PC.InfoLess PC.VarName)
+collectVars' collected (PC.TyVarI tv) = Set.insert (PC.mkInfoLess . view #varName $ tv) collected
+collectVars' collected (PC.TyAppI (PC.TyApp _ args _)) = collected `Set.union` (Set.unions . fmap collectVars $ args)
+collectVars' collected _ = collected
```

Comment:
Is there an implementation for this that you can just import?

Reply:
I actually just copied this from the Haskell implementation, and it worked for our use case